dipy/reconst/ivim.py

```diff
@@ -0,0 +1,383 @@
+        if type(index) is not tuple:
+            index = (index,)
+        elif len(index) >= model_params.ndim:
+            raise IndexError("IndexError: invalid index")
+        index = index + (slice(None),) * (N - len(index))
+        return type(self)(self.model, model_params[index])
+
+    @property
+    def shape(self):
+        return self.model_params.shape[:-1]
+
+    def predict(self, gtab, step=None, S0=1.0):
+        r"""
+        Given a model fit, predict the signal.
+
+        Parameters
+        ----------
+        gtab : a GradientTable class instance
+            This encodes the directions for which a prediction is made
+
+        Notes
+        -----
+        The predicted signal is given by:
+
+        .. math ::
+
+
+
+        Where:
+        .. math ::
+
+        """
+        return ivim_function(self.model_params, gtab.bvals)
+
+
+def one_stage(data, gtab, x0, jac, bounds, tol, routine, algorithm,
+              gtol, ftol, eps, maxiter):
+    """
+    Fit the ivim params using minimize
+
+    Parameters
+    ----------
+    data : array ([X, Y, Z, ...], g)
+        Data or response variables holding the data. Note that the last
+        dimension should contain the data. It makes no copies of data.
+
+    jac : bool
+        Use the Jacobian? Default: False
+
+    Returns
+    -------
+    ivim_params: the S0, f, D_star, D value for each voxel.
+
+    """
+    flat_data = data.reshape((-1, data.shape[-1]))
+    flat_x0 = x0.reshape(-1, x0.shape[-1])
+    # Flatten for the iteration over voxels:
+    bvals = gtab.bvals
+    ivim_params = np.empty((flat_data.shape[0], 4))
+    flat_x0[..., 0] = flat_data[..., 0]
+
+    if routine == 'minimize':
+        result = _minimize(flat_data, bvals, flat_x0, ivim_params,
+                           bounds, tol, jac, algorithm, gtol, ftol, eps, maxiter)
+    elif routine == "leastsq":
+        result = _leastsq(flat_data, bvals, flat_x0, ivim_params)
+    ivim_params.shape = data.shape[:-1] + (4,)
+    return ivim_params
+
+
+def _minimize(flat_data, bvals, flat_x0, ivim_params,
+              bounds, tol, jac, algorithm, gtol, ftol, eps, maxiter):
+    """Use minimize for finding ivim_params"""
+    sum_sq = lambda params, bvals, signal: np.sum(
+        _ivim_error(params, bvals, signal)**2)
+
+    result = []
+    if jac == True:
+        jacobian = _ivim_jacobian_func
+    else:
+        jacobian = None
+    for vox in range(flat_data.shape[0]):
+        res = Optimizer(sum_sq,
+                        flat_x0[vox],
+                        args=(bvals, flat_data[vox]), bounds=bounds,
+                        tol=tol, method=algorithm, jac=jacobian,
+                        options={'gtol': gtol, 'ftol': ftol, 'eps': eps, 'maxiter': maxiter})
+        ivim_params[vox, :4] = res.xopt
+        result += [res]
+    return result
+
+
+def _leastsq(flat_data, bvals, flat_x0, ivim_params):
+    """Use minimize for finding ivim_params"""
+    result = []
```

Comment:
I'm pretty sure we can preallocate this in advance as an `empty` or `zeros` array: 

    result = np.zeros(flat_data.shape[0] + (4,))